File: Sample/Engine/Animation.py
# -*- encoding: utf-8 -*-

from __future__ import annotations
import os
import math
import zlib
from typing import List, Optional, Dict, Any
from . import Sprite, Texture, Image, C_CompressAnimation


def _resolveSoundStopAtEndFrame(entry: Dict[str, Any], frameRate: int) -> None:
    if entry.get("stopAtEndFrame") is not None:
        return
    originalDuration = entry.get("originalDuration")
    if not isinstance(originalDuration, (int, float)) or float(originalDuration) <= 0.0:
        entry["stopAtEndFrame"] = False
        return
    startFrame = int(entry.get("startFrame", 0) or 0)
    endFrame = int(entry.get("endFrame", -1) or -1)
    if endFrame < 0 or frameRate <= 0:
        entry["stopAtEndFrame"] = False
        return
    effectiveDuration = (endFrame - startFrame) / frameRate
    entry["stopAtEndFrame"] = effectiveDuration + 1e-5 < float(originalDuration)


def getAnimationVisualDuration(animationData: Dict[str, Any]) -> float:
    r"""
    \brief Get animation duration based on visual frame segments only.

    - \param animationData Raw or compressed animation data dictionary.
    - \return Visual duration in seconds, excluding sound track length.
    """
    visualDuration = animationData.get("visualDuration")
    if visualDuration is not None:
        return float(visualDuration)
    visualFrameCount = animationData.get("visualFrameCount")
    frameRate = int(animationData.get("frameRate", 30) or 30)
    if frameRate <= 0:
        frameRate = 30
    if visualFrameCount is not None and frameRate > 0:
        return float(visualFrameCount) / frameRate
    if animationData.get("type") == "animation":
        visualMaxTime = 0.0
        for timeLine in animationData.get("timeLines", []):
            for segment in timeLine.get("timeSegments", []):
                if segment.get("type") != "frame":
                    continue
                endTime = segment.get("endFrame", {}).get("time", 0.0)
                if endTime > visualMaxTime:
                    visualMaxTime = endTime
        if visualMaxTime > 0.0:
            return max(1, int(math.ceil(visualMaxTime * frameRate))) / frameRate
        return 0.0
    duration = animationData.get("duration")
    if duration is not None:
        return float(duration)
    frameCount = int(animationData.get("frameCount", 0) or 0)
    if frameCount > 0 and frameRate > 0:
        return float(frameCount) / frameRate
    return 0.0


class AnimSprite(Sprite):
    r"""
    \brief Sprite that plays frame-based animations from compressed data.

    Supports frame rates, frame counts, and embedded sound triggers.
    """

    def __init__(self, animationData: Dict[str, Any]) -> None:
        r"""
        \brief Construct an AnimSprite from animation data.

        - \param animationData Dictionary containing frames, frameRate, frameCount, and sounds.
        """
        self._texture: Texture = None
        self._frames: List[Any] = []
        self._frameRate: int = 30
        self._frameCount: int = 0
        self._frameCounter: float = 0.0
        self._frameIndex: int = 0
        self._finished: bool = False
        self._duration: float = 0.0
        self._visualDuration: float = 0.0
        self._spriteReady: bool = False
        self.setData(animationData)
        super().__init__(self._texture)
        self._spriteReady = True

    def setData(self, animationData: Dict[str, Any]) -> None:
        r"""
        \brief Load animation data and reset playback state.

        - \param animationData Dictionary with frames, frameRate, frameCount, and sounds.
        """
        self._frames = animationData.get("frames", [])
        self._frameRate = int(animationData.get("frameRate", 30) or 30)
        if self._frameRate <= 0:
            self._frameRate = 30
        self._frameCount = int(animationData.get("frameCount", len(self._frames) or 0))
        if self._frameCount <= 0:
            self._frameCount = len(self._frames)
        self.soundEntries = list(animationData.get("sounds", []))
        self.soundEntries.sort(key=lambda entry: int(entry.get("startFrame", 0) or 0))
        for entry in self.soundEntries:
            _resolveSoundStopAtEndFrame(entry, self._frameRate)
        self.soundIndex = 0
        self.playingSounds = []
        self._frameCounter = 0.0
        self._frameIndex = 0
        self._finished = self._frameCount <= 0 or len(self._frames) == 0
        self._duration = float(animationData.get("duration", 0.0) or 0.0)
        if self._duration <= 0.0 and self._frameRate > 0 and self._frameCount > 0:
            self._duration = float(self._frameCount) / self._frameRate
        self._visualDuration = getAnimationVisualDuration(animationData)
        if not self._finished:
            self.applyFrame(0)

    def getDuration(self) -> float:
        r"""
        \brief Get the full animation duration including sound track length.

        - \return Duration in seconds.
        """
        return self._duration

    def getVisualDuration(self) -> float:
        r"""
        \brief Get the visual duration based on frame segments only.

        - \return Visual duration in seconds, excluding sound track length.
        """
        return self._visualDuration

    def isFinished(self) -> bool:
        r"""
        \brief Check whether the animation has played through all frames.

        - \return True if the animation has finished, False otherwise.
        """
        return self._finished

    def getFrameIndex(self) -> int:
        r"""
        \brief Get the current frame index.

        - \return Current frame index (0-based).
        """
        return self._frameIndex

    def update(self, deltaTime: float) -> None:
        r"""
        \brief Advance the animation by the elapsed time and apply the new frame.

        - \param deltaTime Time elapsed since last update (in seconds).
        """
        if self._finished:
            return
        if self._frameCount <= 0:
            self._finished = True
            return
        if deltaTime > 0:
            self._frameCounter += deltaTime * self._frameRate
        newFrameIndex = int(self._frameCounter)
        if newFrameIndex >= self._frameCount:
            self._finished = True
            newFrameIndex = self._frameCount - 1
        if newFrameIndex != self._frameIndex:
            self._frameIndex = newFrameIndex
            self.applyFrame(self._frameIndex)

    def applyFrame(self, frameIndex: int) -> None:
        r"""
        \brief Apply the given frame index to the sprite texture.

        - \param frameIndex Index of the frame to apply.
        """
        if frameIndex < 0 or frameIndex >= len(self._frames):
            return
        frameData = self._frames[frameIndex]
        if not frameData:
            return
        if isinstance(frameData, Image):
            image = frameData
        else:
            memoryData = zlib.decompress(frameData)
            image = Image()
            if not image.loadFromMemory(memoryData, len(memoryData)):
                return
            self._frames[frameIndex] = image
        if self._texture is None:
            self._texture = Texture(image)
            if self._spriteReady:
                self.setTexture(self._texture, True)
            return
        self._texture.update(image)


def compressAnimation(
    animationData: Dict[str, Any], assetsRoot: Optional[str] = None, imageFormat: Optional[str] = "png"
) -> Dict[str, Any]:
    r"""
    \brief Compress animation data into a frame-based format.

    - \param animationData Raw animation data with timelines and assets.
    - \param assetsRoot Root directory for asset resolution (default: "Assets/Animations").
    - \param imageFormat Image format for frame compression (default: "png").
    - \return Compressed animation dictionary with frames and metadata.
    """
    if not animationData:
        return {
            "type": "compressedAnimation",
            "frameRate": 0,
            "frameCount": 0,
            "visualFrameCount": 0,
            "duration": 0.0,
            "visualDuration": 0.0,
            "frames": [],
            "sounds": [],
        }

    frameRate = int(animationData.get("frameRate", 30) or 30)
    if frameRate <= 0:
        frameRate = 30
    frameStep = 1.0 / frameRate

    timeLines: List[Dict[str, List[Dict[str, Any]]]] = animationData.get("timeLines", [])
    assets: List[str] = animationData.get("assets", [])

    maxTime = 0.0
    visualMaxTime = 0.0
    for timeLine in timeLines:
        for segment in timeLine.get("timeSegments", []):
            endTime = segment.get("endFrame", {}).get("time", 0.0)
            if endTime > maxTime:
                maxTime = endTime
            if segment.get("type") == "frame" and endTime > visualMaxTime:
                visualMaxTime = endTime

    frameCount = max(1, int(math.ceil(maxTime * frameRate)))
    if visualMaxTime > 0.0:
        visualFrameCount = max(1, int(math.ceil(visualMaxTime * frameRate)))
        visualDuration = float(visualFrameCount) / frameRate
    else:
        visualFrameCount = 0
        visualDuration = 0.0

    if assetsRoot is None:
        assetsRoot = os.path.join("Assets", "Animations")

    duration, frames, sounds = C_CompressAnimation(
        zlib, frameCount, frameStep, frameRate, timeLines, assets, assetsRoot, imageFormat
    )

    return {
        "type": "compressedAnimation",
        "name": animationData.get("name", ""),
        "frameRate": frameRate,
        "frameCount": frameCount,
        "visualFrameCount": visualFrameCount,
        "duration": duration,
        "visualDuration": visualDuration,
        "frames": frames,
        "sounds": sounds,
    }

File: Sample/Core/Engine/src/Animation.cpp
#include <Animation.hpp>

#include <Base64.hpp>

#include <SFML/Graphics/Color.hpp>
#include <SFML/Graphics/Image.hpp>
#include <SFML/Graphics/Texture.hpp>
#include <zlib.h>

#include <algorithm>
#include <array>
#include <cmath>
#include <cstdint>
#include <stdexcept>
#include <utility>

namespace {
std::unique_ptr<sf::Texture>& placeholderTextureStorage() {
    static std::unique_ptr<sf::Texture> texture;
    return texture;
}

const sf::Texture& placeholderTexture() {
    std::unique_ptr<sf::Texture>& texture = placeholderTextureStorage();
    if (texture == nullptr) {
        texture = std::make_unique<sf::Texture>(
            sf::Image({1u, 1u}, sf::Color::Transparent));
    }
    return *texture;
}

std::vector<std::uint8_t> decodeBase64(const std::string& data) {
    try {
        return ludork::standard::decodeBase64(data);
    } catch (const std::invalid_argument&) {
        throw std::invalid_argument("Invalid base64 animation frame");
    }
}

std::vector<std::uint8_t> decompressFrame(const std::uint8_t* data,
                                          std::size_t size) {
    z_stream stream{};
    stream.next_in = const_cast<Bytef*>(data);
    stream.avail_in = static_cast<uInt>(size);
    if (inflateInit(&stream) != Z_OK) {
        throw std::runtime_error(
            "Failed to initialise animation decompression");
    }
    std::vector<std::uint8_t> output;
    std::array<std::uint8_t, 16384> chunk{};
    int status = Z_OK;
    while (status == Z_OK) {
        stream.next_out = chunk.data();
        stream.avail_out = static_cast<uInt>(chunk.size());
        status = inflate(&stream, Z_NO_FLUSH);
        const std::size_t produced = chunk.size() - stream.avail_out;
        output.insert(output.end(), chunk.begin(), chunk.begin() + produced);
    }
    inflateEnd(&stream);
    if (status != Z_STREAM_END) {
        throw std::runtime_error("Failed to decompress animation frame");
    }
    return output;
}

void resolveSoundStopAtEndFrame(AnimationSoundEntry& entry, int frameRate) {
    if (entry.stopAtEndFrame.has_value()) {
        return;
    }
    if (!entry.originalDuration.has_value() ||
        *entry.originalDuration <= 0.0f || entry.endFrame < 0 ||
        frameRate <= 0) {
        entry.stopAtEndFrame = false;
        return;
    }
    const float effectiveDuration =
        static_cast<float>(entry.endFrame - entry.startFrame) / frameRate;
    entry.stopAtEndFrame = effectiveDuration + 1e-5f < *entry.originalDuration;
}

float getCompressedAnimationVisualDuration(const AnimationData& animationData) {
    if (animationData.visualDuration.has_value()) {
        return *animationData.visualDuration;
    }
    const int frameRate =
        animationData.frameRate > 0 ? animationData.frameRate : 30;
    if (animationData.visualFrameCount.has_value()) {
        return static_cast<float>(*animationData.visualFrameCount) / frameRate;
    }
    if (animationData.duration.has_value()) {
        return *animationData.duration;
    }
    if (animationData.frameCount > 0) {
        return static_cast<float>(animationData.frameCount) / frameRate;
    }
    return 0.0f;
}
}  // namespace

AnimSprite::AnimSprite(const AnimationData& animationData)
    : sf::Sprite(placeholderTexture()) {
    setData(animationData);
}

void AnimSprite::setData(const AnimationData& animationData) {
    frames_ = animationData.frames;
    frameEncoding_ = animationData.frameEncoding.empty()
                         ? "zlib"
                         : animationData.frameEncoding;
    frameRate_ = animationData.frameRate > 0 ? animationData.frameRate : 30;
    frameCount_ = animationData.frameCount > 0
                      ? animationData.frameCount
                      : static_cast<int>(frames_.size());
    if (frameCount_ <= 0) {
        frameCount_ = static_cast<int>(frames_.size());
    }
    soundEntries_ = animationData.sounds;
    timeTags_ = animationData.timeTags;
    std::stable_sort(
        timeTags_.begin(), timeTags_.end(),
        [](const AnimationTimeTag& left, const AnimationTimeTag& right) {
            return left.time < right.time;
        });
    std::stable_sort(
        soundEntries_.begin(), soundEntries_.end(),
        [](const AnimationSoundEntry& left, const AnimationSoundEntry& right) {
            return left.startFrame < right.startFrame;
        });
    for (AnimationSoundEntry& entry : soundEntries_) {
        resolveSoundStopAtEndFrame(entry, frameRate_);
    }
    frameCounter_ = 0.0f;
    frameIndex_ = 0;
    finished_ = frameCount_ <= 0 || frames_.empty();
    duration_ = animationData.duration.value_or(0.0f);
    if (duration_ <= 0.0f && frameRate_ > 0 && frameCount_ > 0) {
        duration_ = static_cast<float>(frameCount_) / frameRate_;
    }
    visualDuration_ = getCompressedAnimationVisualDuration(animationData);
    if (!finished_) {
        applyFrame(0);
    }
}

float AnimSprite::getDuration() const {
    return duration_;
}

float AnimSprite::getVisualDuration() const {
    return visualDuration_;
}

std::vector<AnimationTimeTag> AnimSprite::getAllTimeTags() const {
    return timeTags_;
}

bool AnimSprite::isFinished() const {
    return finished_;
}

int AnimSprite::getFrameIndex() const {
    return frameIndex_;
}

void AnimSprite::update(float deltaTime) {
    if (finished_) {
        return;
    }
    if (frameCount_ <= 0) {
        finished_ = true;
        return;
    }
    if (deltaTime > 0.0f) {
        frameCounter_ += deltaTime * frameRate_;
    }
    int newFrameIndex = static_cast<int>(std::floor(frameCounter_));
    if (newFrameIndex >= frameCount_) {
        finished_ = true;
        newFrameIndex = frameCount_ - 1;
    }
    if (newFrameIndex == frameIndex_) {
        return;
    }
    frameIndex_ = newFrameIndex;
    applyFrame(frameIndex_);
}

void AnimSprite::applyFrame(int frameIndex) {
    if (frameIndex < 0 || frameIndex >= static_cast<int>(frames_.size())) {
        return;
    }
    auto& frame = frames_[static_cast<std::size_t>(frameIndex)];
    std::shared_ptr<sf::Image> image;
    if (std::holds_alternative<std::shared_ptr<sf::Image>>(frame)) {
        image = std::get<std::shared_ptr<sf::Image>>(frame);
    } else {
        const std::string& encoded = std::get<std::string>(frame);
        std::vector<std::uint8_t> compressed;
        if (frameEncoding_ == "base64+zlib") {
            compressed = decodeBase64(encoded);
        } else {
            compressed.assign(encoded.begin(), encoded.end());
        }
        const std::vector<std::uint8_t> memory =
            decompressFrame(compressed.data(), compressed.size());
        image = std::make_shared<sf::Image>();
        if (!image->loadFromMemory(memory.data(), memory.size())) {
            return;
        }
        frame = image;
    }
    if (!image) {
        return;
    }
    if (!texture_) {
        texture_ = std::make_shared<sf::Texture>(*image);
        sf::Sprite::setTexture(*texture_, true);
    } else {
        texture_->update(*image);
    }
}

const std::vector<AnimationSoundEntry>& AnimSprite::getSoundEntries() const {
    return soundEntries_;
}

std::vector<AnimationSoundEntry>& AnimSprite::getSoundEntries() {
    return soundEntries_;
}

float getAnimationVisualDuration(const AnimationSourceData& animationData) {
    if (animationData.visualDuration.has_value()) {
        return *animationData.visualDuration;
    }
    const int frameRate =
        animationData.frameRate > 0 ? animationData.frameRate : 30;
    if (animationData.visualFrameCount.has_value()) {
        return static_cast<float>(*animationData.visualFrameCount) / frameRate;
    }
    if (animationData.type == "animation") {
        float visualMaxTime = 0.0f;
        for (const AnimationTimeline& timeline : animationData.timeLines) {
            for (const AnimationSegment& segment : timeline.timeSegments) {
                if (segment.type == "frame") {
                    visualMaxTime =
                        std::max(visualMaxTime, segment.endFrame.time);
                }
            }
        }
        if (visualMaxTime <= 0.0f) {
            return 0.0f;
        }
        return static_cast<float>(std::max(
                   1, static_cast<int>(std::ceil(visualMaxTime * frameRate)))) /
               frameRate;
    }
    if (animationData.duration.has_value()) {
        return *animationData.duration;
    }
    if (animationData.frameCount > 0) {
        return static_cast<float>(animationData.frameCount) / frameRate;
    }
    return 0.0f;
}

void shutdownAnimationResources() noexcept {
    placeholderTextureStorage().reset();
}
